api/models.py

```diff
@@ -126,3 +133,16 @@ class TramitacaoEvent(models.Model):
 
     class Meta:
         ordering = ('sequencia',)
+
+class EnergiaRecentePeriodo(models.Model):
```

Comment:
Seria bom colocar `help_text` para os atributos dessa classe, alguns não ficaram óbvios para mim.
Proposicao tem um atributo `energia`, é para deixar lá também?

Reply:
Esse atributo é o que o front usa atualmente pra fazer a ordenação de proposições com mais energia e menos energia recentes. No front atualmente, acho que iria se querer era conseguir pegar a energia mais recente das proposições dos últimos X dias(Se não tiver nenhuma energia, seria 0).

Então esse atributo de `energia` tá desatualizado, tava pensando que esse campo poderia ser uma função que varre as energias e verifica a mais recente e me retorna o inteiro. Mas era bom que eu pudesse mudar o período de dias que eu quero... Seria melhor uma endpoint que faria essa operação?

Reply:
ou o front poderia se virar com a endpoint `/energia/<casa>/<id>?periodo=<dias>?` existente e fazer a filtragem por si só pra pegar esse inteiro e a gente tiraria esse atributo `energia`

Reply:
Adicionei os help_texts

Reply:
O atributo energia que está em proposição é o valor da energia atual da proposição (um único valor). Já a energia recente é o histórico da energia acumulada.

Reply:
@danielrcoura eu tava com dúvida nisso, pq tá sendo pego assim no script
`energia_value <- historico_energia[1,]$energia_recente`
e vendo a olho pra todas as energias no json, parece q ele tá pegando primeiro da lista do histórico, q na vdd é o mais antigo, mas n sei se é isso msm pq n sei como tá no R

Reply:
Entendi que Nazareno defendeu usarmos um endpoint mesmo, e o front solicitaria a energia de cada uma.

Reply:
Uma endpoint pra pegar a última energia então?

Reply:
> O atributo energia que está em proposição é o valor da energia atual da proposição (um único valor). Já a energia recente é o histórico da energia acumulada.
Não sei entendi... A energia atual não seria um dos registros da tabela Energia? Pq me parece que seria melhor não ter o atributo energia na tabela Proposicao. A não ser que seja um conceito diferente mesmo.

> Uma endpoint pra pegar a última energia então?
Foi o que entendi do que ele disse. Apesar de que seria mais eficiente se viesse tudo de uma vez só do BD e no endpoint...

Reply:
Opa! Vou me juntar à discussão aqui.

A energia atual deve ser o último valor de energia (referente ao dia atual). Pode ser extraído da tabela de energia mesmo. Só coloquei na proposição pra não quebrar a forma como tava funcionando atualmente.

Sobre essa questão dos links que Nazareno tinha falado, não entendi exatamente o que ele quis dizer. Você já fez algo do tipo antes Andres?

Reply:
Também não sei se a forma como a energia está sendo extraída está correta no backend. Me baseei no código usado pra mostrar o resumo da tramitação. Talvez haja uma maneira mais eficiente de fazer isso.

Reply:
> Sobre essa questão dos links que Nazareno tinha falado, não entendi exatamente o que ele quis dizer. Você já fez algo do tipo antes Andres?

Quando você fala de links entendo que está se referindo a outra coisa que ele disse, diferente do que eu comentei na minha mensagem anterior.

Os links que ele falou para recursos REST entendo que são algo nesse sentido:
https://stormpath.com/blog/linking-and-resource-expansion-rest-api-tips
Em outras palavras, vc não manda todos os objetos relacionados ao que está retornando nessa chamada, mas manda links para que a pessoa possa facilmente acessar esses outros objetos relacionados.

O que eu estava falando na minha mensagem anterior era sobre retornar tudo de uma vez na mesma requisição ou não. A sugestão de Nazareno é não retornar tudo de uma vez. Seria assim:

1- Cliente solicita uma lista de proposições. Retorno é uma lista de objetos (proposições), cada um com informações gerais sobre a proposição e links para "detalhamento" e "energia", por exemplo.
2- Para cada proposição o cliente faz uma nova requisição pedindo a energia da proposição.

**Temos assim N+1 requisições e N+1 queries ao BD.**

Uma outra alternativa é, tendo em mente que já sabemos que o cliente vai precisar mostrar informações gerais e a energia de uma proposição, nosso endpoint que lista as proposições já inclui em cada proposição o valor da energia dela também.

**Temos assim 1 requisição e 1 query ao BD.**

Uma terceira alternativa meio termo é ter um endpoint que retorna a energia de múltiplas proposições de uma vez.

1 - cliente pede lista de proposições
2 - cliente faz uma requisição pedindo a energia das N proposições que recebeu anteriormente. Retorno é uma lista de energias identificadas pelos IDs das proposições.

**Temos assim 2 requisições e 2 queries.**

Ou seja, as duas últimas opções seriam mais eficientes, mas menos *RESTfull* e também talvez possam ser consideradas ["otimizações prematuras"](http://wiki.c2.com/?PrematureOptimization)...

Reply:
Tirei a função de dentro do models, pq pelas documentações não é o mais adequado. O adequado seria trabalhar com queries nas Views

Pra filtrar as energias por período de dias ficou mais enxuto com filter [aqui](https://github.com/analytics-ufcg/agora-digital-backend/blob/8022c52ee43bfe4949d2db72318ed6787c9cc995/api/views.py#L68)

A endpoint pra retornar a última energia tbm foi usando [isso](https://github.com/analytics-ufcg/agora-digital-backend/blob/8022c52ee43bfe4949d2db72318ed6787c9cc995/api/views.py#L103)

Agora models não precisa ter nenhuma função relacionada a energia, acho que fica melhor assim, o que acham?

Reply:
Deem *migrate* e *import_data* pq eu mudei os nomes dos Models pq tava confuso a nomenclatura do histórico e o script do R.

As endpoints ficaram:

- `http://0.0.0.0:8000/energia/<casa>/<id>` pra todas as energias de uma prop
- `http://0.0.0.0:8000/energia/<casa>/<id>?periodo=<dias>` por um periodo de dias
- `http://0.0.0.0:8000/proposicoes/senado/91341/energia` ultima energia_recente

Reply:
Talvez o padrão das URLs não esteja ideal, se vcs tiverem sugestão de como elas ficariam

Reply:
Será que a funcionalidade da última URL não poderia estar inclusa na segunda?

Reply:
Acho que nunca usei esse `get_queryset`, mas me parece uma boa sim.